std/exception.d

```diff
@@ -1810,51 +1807,48 @@ if (isInputRange!IRange)
 
         static if (hasSlicing!IRange)
         {
-            static if (hasLength!IRange)
+            static if (primitivesToHandle & RangePrimitive.opSlice)
             {
-                static if (functionsToHandle & RangePrimitive.opSlice)
+                static if (hasLength!IRange)
                 {
-                    auto opSlice(size_t low, size_t high)
+                    typeof(this) opSlice(size_t lower, size_t upper)
                     {
                         try
                         {
-                            return typeof(this)(this.range[low .. high]);
+                            return typeof(this)(this.range[lower .. upper]);
                         }
                         catch(E exception)
                         {
-                            return handler(exception, this.range);
+                            return typeof(this)(handler(exception, this.range));
                         }
                     }
                 }
-            }
-            else static if (is(typeof(range[opSlice_t.max .. $])))
-            {
-                struct DollarToken{}
-                enum opDollar = DollarToken.init;
-                static if (functionsToHandle & RangePrimitive.opSlice)
+                else static if (is(typeof(IRange.init[size_t.init .. $])))
                 {
-                    auto opSlice(size_t low, DollarToken)
+                    static struct DollarToken {}
+                    enum opDollar = DollarToken.init;
+
+                    typeof(this) opSlice(size_t lower, DollarToken)
                     {
                         try
                         {
-                            return typeof(this)(this.range[low .. $]);
+                            return typeof(this)(this.range[lower .. $]);
                         }
                         catch(E exception)
                         {
-                            return handler(exception, this.range);
+                            return typeof(this)(handler(exception, this.range));
                         }
                     }
 
-                    auto opSlice(size_t low, size_t high)
+                    auto opSlice(size_t lower, size_t upper)
                     {
                         try
                         {
-                            import std.range : take;
-                            return this[low .. $].take(high - low);
+                            return takeExactly(typeof(this)(this.range[lower .. $]), upper - 1);
                         }
                         catch(E exception)
                         {
-                            return handler(exception, this.range);
+                            return takeExactly(typeof(this)(handler(exception, this.range)), 0);
```

Comment:
it only has return Take if it is a infinite range. Otherwise it has to return typeof(this).
IMO the handler has to return the correct thing. any empty Take is properly worse than an exception (silent error)